### ISSUE TITLE
Adding Retry Logic into the warehouse

### ISSUE THREAD

**Issue by ramiroJCB** [mergestat/mergestat#408]
With the addition of github actions we saw the need of retry logic .For this we will need to :

- [x] Identify which type of calls are worth  retrying (thinking of  saving quota)
- [x] If a call is worth retrying , how many times should we retry it (Idempotence) ?
- [x] how much time we should wait to retry  per call
- [ ] Initially we should implement a basic retry logic with the bullets  above , but should  build on top of that retry better patterns
- [ ]  this retry  logic will be implemented for the moment only the package warehouse  .

**Comment by ramiroJCB** [comment]
so @patrickdevivo  @amenowanna  , with lastest discoveries with debug logs and retries .I've found :

1. most of common errors from the github api are volatile  meaning that it may be initially be a 500 but if you retry enough it changes to 410 status gone , if you keep retrying  on it does not show signals from recovery or getting the actual content of the call  , I have not experienced  any retry either externally or locally that recovers from the unexpected. status .
2. most if not all calls are not worth retrying  , I have not seen the only error that I've suggested as retryable that is 503 to experience the retry  behavior on it . 
3. with such  consistent error prone  behavior from the github api ,I will say that we should not fail on any error from the github api and the retry logic should only applied  for 503 with max of 3 retries  with a max of  1 hour for the exponential backoff to finally move pass it .
4.We should get 1 item per page to reduce the chances of loosing  valid data.
5.We should not  fail on any error from the github api ,Our system should resilience enough to just extract as much data possible from it  each run . With this volatile behabior from  their v3 api  we should log in a debug/warning log the error so  trouebleshoot or to  have  a way to validate that the information is missing from source  if someone request it .
